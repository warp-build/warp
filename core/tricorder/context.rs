use crate::resolver::TargetRegistry;
use crate::sync::Arc;
use crate::worker::TaskRegistry;

#[derive(Debug, Clone)]
pub struct TricorderContext {
    pub target_registry: Arc<TargetRegistry>,
    pub task_registry: Arc<TaskRegistry>,
}

impl TricorderContext {
    pub fn new(target_registry: Arc<TargetRegistry>, task_registry: Arc<TaskRegistry>) -> Self {
        Self {
            target_registry,
            task_registry,
        }
    }
}
