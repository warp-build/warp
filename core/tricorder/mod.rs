mod grpc;
mod manager;
mod registry;

pub use grpc::*;
pub use manager::*;
pub use registry::*;

use crate::archive::Archive;
use crate::model::{
    ConcreteTarget, ExecutableSpec, Requirement, Signature, SignatureError, TargetId,
};
use crate::store::ArtifactManifest;
use crate::sync::*;
use async_trait::async_trait;
use std::fmt::Debug;
use thiserror::*;
use url::Url;

const DEFAULT_TRICODER_BINARY_NAME: &str = "tricorder";

#[derive(Debug, Clone)]
pub struct Connection {
    pub(crate) tricorder_url: Url,
    pub(crate) port: u16,
}

#[async_trait]
pub trait Tricorder: Send + Sync + Debug {
    async fn connect(connection: Connection) -> Result<Self, TricorderError>
    where
        Self: Sized;

    async fn ensure_ready(&mut self) -> Result<(), TricorderError>;

    async fn generate_signature(
        &mut self,
        concrete_target: &ConcreteTarget,
        dependencies: &[(Arc<ExecutableSpec>, Arc<ArtifactManifest>)],
    ) -> Result<SignatureGenerationFlow, TricorderError>;

    async fn ready_dependency(
        &mut self,
        concrete_target: &ConcreteTarget,
        archive: &Archive,
    ) -> Result<SignatureGenerationFlow, TricorderError>;
}

#[derive(Error, Debug)]
pub enum TricorderError {
    #[error("Something went wrong with this tricorder")]
    Unknown,

    #[error("Something went wrong with a gRPC tricorder: {0:?}")]
    GrpcError(tonic::Status),

    #[error(transparent)]
    SignatureError(SignatureError),
}

#[derive(Debug)]
pub enum SignatureGenerationFlow {
    GeneratedSignatures { signatures: Vec<Signature> },
    MissingRequirements { requirements: Vec<Requirement> },
    IgnoredTarget(TargetId),
}
