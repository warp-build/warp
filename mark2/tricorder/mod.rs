mod grpc;
mod manager;
mod registry;

pub use grpc::*;
pub use manager::*;
pub use registry::*;

use crate::model::ConcreteTarget;
use async_trait::async_trait;
use std::fmt::Debug;
use thiserror::*;

#[derive(Error, Debug)]
pub enum TricorderError {
    #[error("Something went wrong with this tricorder")]
    Unknown,

    #[error("Something went wrong with a gRPC tricorder: {0:?}")]
    GrpcError(tonic::Status),
}

pub enum SignatureGenerationFlow {
    //GeneratedSignatures { signatures: Vec<Signature> },
    //MissingRequirements { requirements: Vec<Requirement> },
}

const DEFAULT_TRICODER_BINARY_NAME: &str = "tricorder";

#[derive(Debug, Clone, Copy)]
pub struct Connection {
    port: u16,
}

#[async_trait]
pub trait Tricorder: Send + Sync + Debug {
    async fn connect(connection: Connection) -> Result<Self, TricorderError>
    where
        Self: Sized;

    async fn ensure_ready(&self) -> Result<(), TricorderError>;

    async fn generate_signature(
        &self,
        concrete_target: &ConcreteTarget,
    ) -> Result<SignatureGenerationFlow, TricorderError>;
}
