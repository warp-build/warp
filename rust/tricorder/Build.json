{
    "signatures": [
	{
	    "name": "tricoder",
	    "rule": "cargo_binary",
	    "bin": "tricoder",
	    "srcs": ["**/*"]
	}
    ]
}
