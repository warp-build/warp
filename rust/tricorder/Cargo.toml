[package]
name = "tricorder"
version = "0.1.0"
edition = "2021"

[dependencies]
anyhow = "1.0"
deno_core = "0.154"
thiserror = "1.0"
tonic = "0.8.2"
tokio = { version = "1", features = ["full"] }
tracing = { version = "0.1", features = ["log"] }
tracing-opentelemetry = "0.17"
tracing-subscriber = "0.3"
prost = "0.11.2"

[[bin]]
name = "rust-tricorder"
path = "src/main.rs"

[build-dependencies]
tonic-build = "0.8.2"
