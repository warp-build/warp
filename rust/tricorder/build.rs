use std::path::PathBuf;

fn main() {
    tonic_build::configure()
        .build_server(true)
        .include_file("_include.rs")
        .compile_well_known_types(true)
        .compile(
            &[
                "../../schemas/build/warp/codedb/analyzer.proto",
                "../../schemas/build/warp/resolver.proto",
                "../../schemas/build/warp/common.proto",
            ],
            &["../.."],
        )
        .unwrap_or_else(|e| panic!("protobuf compile error: {}", e));

    let dir: PathBuf = ["vendor/tree-sitter-rust", "src"].iter().collect();

    cc::Build::new()
        .include(&dir)
        .file(dir.join("parser.c"))
        .file(dir.join("scanner.c"))
        .warnings(false)
        .compile("tree-sitter-rust");
}
