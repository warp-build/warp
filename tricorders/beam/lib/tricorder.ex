defmodule Tricorder do
  @moduledoc """
  Documentation for `Tricorder`.
  """

  @doc """
  Hello world.

  ## Examples

      iex> Tricorder.hello()
      :world

  """
  def hello do
    :world
  end
end
