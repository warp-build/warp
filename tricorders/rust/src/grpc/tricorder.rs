use super::mappers::ast_to_success_response;
use super::proto::build::warp::tricorder::tricorder_service_server::{
    TricorderService, TricorderServiceServer,
};
use super::proto::build::warp::tricorder::{
    generate_signature_response, EnsureReadyRequest, EnsureReadyResponse, GenerateSignatureRequest,
    GenerateSignatureResponse, GenerateSignatureSuccessResponse, GetAstRequest, GetAstResponse,
    PrepareDependencyRequest, PrepareDependencyResponse,
};
use super::GrpcTricorderError;
use crate::analysis::model::{Ast, Symbol};
use crate::analysis::Analyzer;
use crate::dependencies::DependencyManager;
use std::net::SocketAddr;
use std::path::Path;
use std::sync::Arc;
use tonic::transport::Server;
use tonic::{Request, Response, Status};

#[derive(Builder)]
#[builder(build_fn(error = "GrpcTricorderError"))]
pub struct GrpcTricorder {
    address: SocketAddr,
    dep_manager: Arc<DependencyManager>,
    analyzer: Arc<Analyzer>,
}

impl GrpcTricorder {
    pub fn builder() -> GrpcTricorderBuilder {
        Default::default()
    }

    pub async fn serve(self) -> Result<(), GrpcTricorderError> {
        let address = self.address;

        Server::builder()
            .add_service(TricorderServiceServer::new(self))
            .serve(address)
            .await?;

        Ok(())
    }
}

#[tonic::async_trait]
impl TricorderService for GrpcTricorder {
    async fn ensure_ready(
        &self,
        _request: Request<EnsureReadyRequest>,
    ) -> Result<Response<EnsureReadyResponse>, Status> {
        self.dep_manager.wait_until_ready().await;
        Ok(Response::new(EnsureReadyResponse::default()))
    }

    async fn generate_signature(
        &self,
        request: Request<GenerateSignatureRequest>,
    ) -> Result<Response<GenerateSignatureResponse>, Status> {
        let request_data = request.into_inner();
        let file = request_data.clone().file;
        let test_matcher: Vec<String> = match request_data.test_matcher {
            Some(matches) => matches.into(),
            None => Vec::new(),
        };
        let workspace_root = request_data.workspace_root;

        let response = self
            .analyzer
            .generate_signature(workspace_root.clone(), Path::new(&file), test_matcher)
            .await;

        match response {
            Ok(signature) => Ok(Response::new(GenerateSignatureResponse {
                response: Some(generate_signature_response::Response::Ok(
                    GenerateSignatureSuccessResponse {
                        workspace_root,
                        file,
                        signatures: signature.iter().map(|e| e.clone().into()).collect(),
                    },
                )),
            })),
            // TODO(@calin): create a proper way of dealing with missing dependencies.
            Err(_) => Ok(Response::new(GenerateSignatureResponse {
                response: Some(generate_signature_response::Response::Ok(
                    GenerateSignatureSuccessResponse {
                        workspace_root,
                        file,
                        signatures: vec![],
                    },
                )),
            })),
        }
    }

    async fn get_ast(
        &self,
        request: Request<GetAstRequest>,
    ) -> Result<Response<GetAstResponse>, Status> {
        let request = request.into_inner();

        let sym = match request.test_matcher {
            Some(test_matcher) if !test_matcher.raw.is_empty() => Symbol::Named {
                name: test_matcher.raw[0].to_string(),
            },
            _ => Symbol::All,
        };

        let response = self.analyzer.get_ast(request.file, sym).await;

        match response {
            Ok(ast) => Ok(Response::new(ast_to_success_response(ast))),
            // TODO(@calin): create a proper Requirement internal model for missing deps
            // and return it properly here.
            Err(_) => Ok(Response::new(ast_to_success_response(Ast::default()))),
        }
    }

    async fn prepare_dependency(
        &self,
        _request: Request<PrepareDependencyRequest>,
    ) -> Result<Response<PrepareDependencyResponse>, Status> {
        Ok(Response::new(PrepareDependencyResponse::default()))
    }
}
