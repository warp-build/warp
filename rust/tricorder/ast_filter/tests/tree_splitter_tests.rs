use tricorder::TreeSplitter;

#[test]
fn t001_simple_fn_graph() {
    let src = include_str!("./sources/t001_simple_fn_graph.rs");
    let symbol = "potato";
    let ast = TreeSplitter::tree_split(symbol, src);
    insta::assert_debug_snapshot!(ast);
}
