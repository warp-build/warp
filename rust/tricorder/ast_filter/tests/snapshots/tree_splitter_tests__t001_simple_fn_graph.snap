---
source: tests/tree_splitter_tests.rs
expression: ast
---
File {
    shebang: None,
    attrs: [],
    items: [
        Fn(
            ItemFn {
                attrs: [],
                vis: Inherited,
                sig: Signature {
                    constness: None,
                    asyncness: None,
                    unsafety: None,
                    abi: None,
                    fn_token: Fn,
                    ident: Ident(
                        potato,
                    ),
                    generics: Generics {
                        lt_token: None,
                        params: [],
                        gt_token: None,
                        where_clause: None,
                    },
                    paren_token: Paren,
                    inputs: [],
                    variadic: None,
                    output: Type(
                        RArrow,
                        Path(
                            TypePath {
                                qself: None,
                                path: Path {
                                    leading_colon: None,
                                    segments: [
                                        PathSegment {
                                            ident: Ident(
                                                bool,
                                            ),
                                            arguments: None,
                                        },
                                    ],
                                },
                            },
                        ),
                    ),
                },
                block: Block {
                    brace_token: Brace,
                    stmts: [
                        Expr(
                            Call(
                                ExprCall {
                                    attrs: [],
                                    func: Path(
                                        ExprPath {
                                            attrs: [],
                                            qself: None,
                                            path: Path {
                                                leading_colon: None,
                                                segments: [
                                                    PathSegment {
                                                        ident: Ident(
                                                            whatever,
                                                        ),
                                                        arguments: None,
                                                    },
                                                ],
                                            },
                                        },
                                    ),
                                    paren_token: Paren,
                                    args: [],
                                },
                            ),
                        ),
                    ],
                },
            },
        ),
    ],
}
